libtransmission/session.cc

```diff
@@ -1211,129 +1196,87 @@ void tr_session::closeImplStart()
             auto const b_cur = b->downloadedCur + b->uploadedCur;
             return a_cur > b_cur; // larger xfers go first
         });
-
     for (auto* tor : torrents)
     {
         tr_torrentFree(tor);
     }
-
     torrents.clear();
-
-    /* Close the announcer *after* closing the torrents
-       so that all the &event=stopped messages will be
-       queued to be sent by tr_announcerClose() */
+    // ...and now that all the torrents have been closed, any
+    // remaining `event=stopped` announce messages are queued in
+    // the announcer. The announcer's destructor sends all those
+    // out via `web_`...
     tr_announcerClose(this);
-
-    /* and this goes *after* announcer close so that
-       it won't be idle until the announce events are sent... */
-    this->web_->closeSoon();
-
+    // ...and now that those are queued, tell web_ that we're
+    // shutting down soon. This leaves the `event=stopped` messages
+    // in the queue but refuses to take any _new_ tasks
+    this->web_->startShutdown();
     this->cache.reset();
 
-    /* saveTimer is not used at this point, reusing for UDP shutdown wait */
+    // recycle the now-unused save_timer_ here to wait for UDP shutdown
     TR_ASSERT(!save_timer_);
-    save_timer_ = timerMaker().create([this]() { closeImplWaitForIdleUdp(); });
-    save_timer_->start(1ms);
+    save_timer_ = timerMaker().create([this]() { closeImplPt2(); });
+    save_timer_->startRepeating(50ms);
 }
 
-void tr_session::closeImplWaitForIdleUdp()
+void tr_session::closeImplPt2()
 {
-    /* gotta keep udp running long enough to send out all
-       the &event=stopped UDP tracker messages */
+    // try to keep the UDP announcer alive long enough to send out
+    // all the &event=stopped tracker announces
     if (announcer_udp_ && !announcer_udp_->isIdle())
     {
         announcer_udp_->upkeep();
-        save_timer_->start(100ms);
         return;
     }
 
-    closeImplFinish();
-}
-
-void tr_session::closeImplFinish()
-{
     save_timer_.reset();
 
-    /* we had to wait until UDP trackers were closed before closing these: */
     this->announcer_udp_.reset();
     this->udp_core_.reset();
 
     stats().saveIfDirty();
     peer_mgr_.reset();
     tr_utpClose(this);
-    blocklists_.clear();
     openFiles().closeAll();
     is_closed_ = true;
 }
 
-static bool deadlineReached(time_t const deadline)
-{
-    return time(nullptr) >= deadline;
-}
-
-static auto constexpr ShutdownMaxSeconds = time_t{ 20 };
-
 void tr_sessionClose(tr_session* session)
 {
     TR_ASSERT(session != nullptr);
 
-    time_t const deadline = time(nullptr) + ShutdownMaxSeconds;
+    static auto constexpr DeadlineSecs = 20s;
```

Comment:
I know it's not the scope of this PR, but on macOS, the system may kill apps that take longer than 10s to close. So can you make a conditional like:
```
#ifdef __APPLE__
static auto constexpr DeadlineSecs = 10s;
#else
static auto constexpr DeadlineSecs = 20s;
#endif
```

It may be especially useful when attempting to restart the mac while the app is open.

Reply:
I'd rather have the behavior consistent -- I'll lower it to 10s everywhere. f41bd601a395fe33ecc32da730ad0446c476529e